core/invoker/src/main/scala/whisk/core/invoker/ActionLogDriver.scala

```diff
@@ -0,0 +1,119 @@
+        implicit transid: TransactionId, logging: Logging): (Boolean, Boolean, Vector[LogLine]) = {
+
+        if (logMsgs.nonEmpty) {
+            val records = logMsgs.split("\n").toStream flatMap {
+                line =>
+                    Try { line.parseJson.convertTo[LogLine] } match {
+                        case Success(t) => Some(t)
+
+                        case Failure(t) =>
+                            // Drop lines that did not parse to JSON objects.
+                            // However, should not happen since we are using the json log driver.
+                            logging.error(this, s"log line skipped/did not parse: $t")
+                            None
+                    }
+            }
+
+            val cumulativeSize = records.scanLeft(0.bytes) { (acc, current) =>
+                acc + {
+                    // do not count the sentinels
+                    if (current.log != LOG_ACTIVATION_SENTINEL) current.log.sizeInBytes else 0.B
+                }
+            }.tail
+
+            val logBuffer = records.zip(cumulativeSize).takeWhile(_._2 <= limit)
+            val isTruncated = logBuffer.size < records.size
+
+            if (requireSentinel) {
+                // there may be more than one sentinel in stdout/stderr (as logs may leak across activations
+                // if for example there log limit was exceeded in one activation and the container was reused
+                // to run the same action again (this is considered a feature - otherwise, must drain the logs
+                // or destroy the container as if it errored)
+                val (sentinels, regulars) = logBuffer.partition(_._1.log.trim == LOG_ACTIVATION_SENTINEL)
+                val hasOut = sentinels.exists(_._1.stream == "stdout")
+                val hasErr = sentinels.exists(_._1.stream == "stderr")
+
+                val bytesRead = if (regulars.nonEmpty) regulars.last._2 else 0.B
+                (hasOut && hasErr, isTruncated, appendTruncateLastLine(regulars, records, logBuffer.size, limit, isTruncated))
+            } else {
+                (true, isTruncated, appendTruncateLastLine(logBuffer, records, logBuffer.size, limit, isTruncated))
+            }
+        } else {
+            logging.warn(this, s"log message is empty")
+            (!requireSentinel, false, Vector())
+        }
+    }
+
+    private def appendTruncateLastLine(resultBuffer: Stream[(LogLine, ByteSize)], allRecords: Stream[LogLine], nextLineIndex: Int, limit: ByteSize, truncate: Boolean) = {
+        if (truncate) {
+            val bytesRead = if (resultBuffer.nonEmpty) resultBuffer.last._2 else 0.B
+            val readPartial = limit - bytesRead
+            if (readPartial > 0.B) {
+                val truncatedLine = allRecords(nextLineIndex).truncateTo(limit - bytesRead)
+                resultBuffer.map(_._1).toVector :+ truncatedLine
+            } else resultBuffer.map(_._1).toVector
+        } else resultBuffer.map(_._1).toVector
```

Comment:
extract `resultBuffer.map(_._1).toVector` to a value up top, i.e. `val logsRead = resultBuffer.map(_._1).toVector`